### PR TITLE
USSD session - graceful termination

### PR DESCRIPTION
**Pull request by nullcode**
I wrote a DFS based algorithm that checks whether a session should be closed at a certain point in a flow. The question is how should I pass that information to the SEND_FUNCTIONS. 

Since I don't want to mess around with new fields on Msg I leave it as an open question for you guys @ericnewcomer @nicpottier. My idea is to pass it as an extra property of the runtime created `MsgStruct` that is pushed to `task_msgs`, but you guys might know better.

### CODE REVIEW

**Comment by ericnewcomer** [comment]
Hey @nullcode, sorry for the delay, this one is a bit sticky. After deliberating we think that the cleanest approach would be to add a nullable session on Msg. We'll go ahead and add that and it'll be pushed out by the end of the week so you can fold that in to complete this PR.

**Comment by nullcode** [comment]
@ericnewcomer that's great, that would be the best for me too.

**Comment by nullcode** [comment]
@ericnewcomer Have you considered m2m to Msg on session instead?

**Comment by ericnewcomer** [comment]
Yup, but if it's a nullable field it shouldn't be any more expensive. As such, it seems we should model it in the most natural way since it is 1-N.

**Comment by smn** [comment]
@ericnewcomer what's the status of this? I'm looking into USSD support for the Junebug channel and that also needs some way of determining whether to set the session flag to `resume` or `end`. Currently there's no signalling for that.

**Comment by ericnewcomer** [comment]
@smn we did add the session_id on msg so @nullcode is free to wrap this up. Sorry neglected to update this ticket when we completed that work.

**Comment by smn** [comment]
thanks @ericnewcomer, @nullcode I see the session is a ForeignKey to `channels.ChannelSession`, how does this relate (if at all) to `ussd.USSDSession` ?

**Comment by ericnewcomer** [comment]
`USSDSession` is a proxy model for `ChannelSession`. So the `session_id` on `Msg` applies to both it and `IVRCall`.

**Comment by smn** [comment]
@nullcode is this on your radar to continue with as part of your original USSD work?

**Comment by nullcode** [comment]
@smn yeah, I will continue working on it from next week

**Comment by smn** [comment]
@nullcode thanks! Shout if there's something we can help with.

**Comment by smn** [comment]
@nullcode have you been able to make any progress on this? We're currently contracted to deliver a USSD based project using RapidPro and need to get a sense of when this will be attended to so as to avoid having two contracts covering essentially the same work.

**Comment by nullcode** [comment]
@smn yeah I started it last week, but since this a complex one I'll need to test all the use cases and edge cases, I'd say 5 more days to complete it...then we can expect some review cycles as well.

File: temba/channels/models.py [hunk 1]
```diff
@@ -1905,16 +1905,27 @@ def send_smscentral_message(cls, channel, msg, text):
     def send_vumi_message(cls, channel, msg, text):
         from temba.msgs.models import WIRED, Msg
         from temba.contacts.models import Contact
+        from temba.ussd.models import USSDSession
 
         is_ussd = channel.channel_type in Channel.USSD_CHANNELS
         channel.config['transport_name'] = 'ussd_transport' if is_ussd else 'mtech_ng_smpp_transport'
```
Comment (nullcode) [inline temba/channels/models.py]: oh okay, kinda figured when I used my local deployment of vumi, but wasn't sure about the sms part in prod. thx

Comment (nullcode) [inline temba/channels/models.py]: we actually have this in the claim form to customize too, I'd make a separate PR for that, because it has no relation to this feature

**Comment by nullcode** [comment]
So this is how it looks like:
- Option to end a USSD session with a message:
screen shot 2017-03-02 at 13 04 03

- If you don't use the action to close the session RapidPro will figure out when to close the session and will do it automatically.

@ericnewcomer @nicpottier review plz

File: temba/channels/models.py (hunk 1 above)
Comment (smn) [inline temba/channels/models.py]: 👌

**Comment by smn** [comment]
@nullcode out of interest, if we're able to algorithmically determine when a session should be ended, why do we need the option to end the "End USSD Session"? Is it simply because explicit is better or because there are scenarios where the algorithmic approach fails?

**Comment by nullcode** [comment]
@smn it's always the algorithmic approach that closes the session, we just needed to introduce a new action for the users so they can specify the "last" aka closing message (this is a special action that cannot be extended further). You can't do that with USSD rulesets only, because they are always waiting for a reply from the contact.

**Comment by nullcode** [comment]
@nicpottier @smn `msgs/models.py:733` & `msgs/models.py:740`, why was USSD added here? This way no messages are actually getting sent.

**Comment by nicpottier** [comment]
Have to ask @smn We don't have any USSD channels to test with so assumed he had played with this some.

**Comment by smn** [comment]
@nullcode hm yeah, looks like this is wrong, I think I'm getting the pull & push USSD mixed up here. Happy to see reverted.

**Comment by smn** [comment]
@nullcode thanks for adding the session end stuff for junebug and moving the tests.

**Comment by nicpottier** [comment]
@nullcode is this ready to review again?

**Comment by nullcode** [comment]
@nicpottier yeah sure!

**Comment by smn** [comment]
Hello! Is there anything else needing to be done here other than the review? We'd like to be able to close our USSD sessions :)

File: temba/channels/models.py [hunk 2]
```diff
@@ -3600,21 +3623,19 @@ class ChannelSession(SmartModel):
...
+                      (ENDING, "Ending"))
 
     external_id = models.CharField(max_length=255,
                                    help_text="The external id for this session, our twilio id usually")
     status = models.CharField(max_length=1, choices=STATUS_CHOICES, default=PENDING,
                               help_text="The status of this session")
-
     channel = models.ForeignKey('Channel',
-                                help_text="The channel that created this session")
+                                help_text="The channel that created this session", null=True)
```
Comment (ericnewcomer) [inline temba/channels/models.py]: Can you explain why channel would be nullable on session?

File: temba/channels/models.py [hunk 3]
```diff
@@ -3628,10 +3649,26 @@ class ChannelSession(SmartModel):
...
+                self.__class__ = USSDSession
+            elif self.session_type == self.IVR:
+                from temba.ivr.models import IVRCall
+                self.__class__ = IVRCall
+
     def get_logs(self):
         return self.channel_logs.all().order_by('created_on')
 
-    def get_duration(self):
+    def get_duration(self):  # pragma: needs cover
```
Comment (ericnewcomer) [inline temba/channels/models.py]: shouldn't add new needs cover pragmas

File: templates/flows/flow_editor.haml [hunk 4]
```diff
@@ -448,8 +448,8 @@
                     [[action.name]]
 
 
-          .add.icon-plus-circle2{ng-click:'addAction(action_set);'}
-          .source{ng-attr-id:'[[action_set.uuid]]_source', drop-scope:'rules', ng-class:'{connected:action_set.destination}', ng-click:'clickActionSource(action_set); $event.stopPropagation()'}
+          .add.icon-plus-circle2{ng-show:'!hasEndUssd(action_set)', ng-click:'addAction(action_set);'}
+          .source{ng-show:'!hasEndUssd(action_set)', ng-attr-id:'[[action_set.uuid]]_source', drop-scope:'rules', ng-class:'{connected:action_set.destination}', ng-click:'clickActionSource(action_set); $event.stopPropagation()'}
```
Comment (ericnewcomer) [inline templates/flows/flow_editor.haml]: Have you accounted for already connected actionsets that have been given a new end ussd action?

File: temba/channels/models.py (hunk 2 above)
Comment (nullcode) [inline temba/channels/models.py]: the scenario is that a user creates a ussd flow without a channel added and wants to simulate the flow

File: temba/channels/models.py (hunk 3 above)
Comment (nullcode) [inline temba/channels/models.py]: I added this for you :) I mean this was your method and coveralls was failing on me because it wasn't covered before (I don't know why it wasn't failing for you when you added it)

File: templates/flows/flow_editor.haml (hunk 4 above)
Comment (nullcode) [inline templates/flows/flow_editor.haml]: thanks for pointing this out, fixed!

**Comment by SeanBlaschke** [comment]
Great to see that it looks like all the issues have been solved. Can we kindly get a review by Monday? We are launching the Uganda FamilyConnect Programme on Tuesday, and this is the final showstopper to be resolved. Thanks for all the work on this!

**Comment by nicpottier** [comment]
@SeanBlaschke That you are at all dependent on us merging this into mainline points to a very broken process. That business owners are pressuring merges points to a broken cultural process.

The entire reason for forks and lightweight branching is to avoid this very thing. Additionally, that this development is taking place on the /rapidpro branch is detrimental to the entire community. The way this is meant to work is for features to be worked on and vetted out on vendor forks, and once tested on production and once the vendor is confident of the changes THEN a merge to /rapidpro happens.

This is how Nyaruka and Ihlasoft works and is the only sane way to do things. The community gets the benefit of having PRs vetted on real installs before they affect everyone and the vendors aren't blocked by anybody and can serve their customers promptly.

I am sure @smn is in agreement on this. We would be much happier to have these changes vetted on the Praekelt branch, tested on a real deploy before having the entire community take them in. @ewheeler

**Comment by SeanBlaschke** [comment]
@nicpottier My comment was based on a specific request yesterday from Praekelt that they required more support directly engaging with partners to quickly resolve this issue, and @ewheeler 's recommendation that all communication be channeled through Github. I have no issues with your guidance on how to test and implement new code, as I am not involved in this part of the process. If this process is broken and these features are being developed on the /rapidpro branch rather than a fork, please provide this specific guidance directly to the vendors and we will reinforce compliance. If this is not the place for business owners to engage on feature issues flagged by our vendors, than please let us know the proper forum as this is what had been recommended.

**Comment by nicpottier** [comment]
Ok great. GitHub is great a place to do the communication, but really you guys need to unblock yourself from being merged into /rapidpro. If Praekelt is saying that is blocking them then they need to change their process, as that shouldn't be the case as that is obviously bad for everyone. (hot fixes would have to be accepted on the branch everybody uses instead of living on a fork until more proper fixes are made etc..)

So my recommendation is to pull this PR into a Praekelt branch, launch it there and get some experience on it then merge back into the master branch. That is the general process we should all be following.

**Comment by SeanBlaschke** [comment]
Thanks @nicpottier. Again, everything you are recommending makes complete sense. @smn - moving forward, what was the rationale why we didn't work on this on a fork first? Is it too late to do this so as not to hold up the work?

**Comment by smn** [comment]
@SeanBlaschke & @nicpottier Thanks all for picking this up. We weren't contracted to do this USSD work and the fork isn't ours.

If this is ready to be QA tested then we can take this code, test it on the Uganda QA environment to see how it performs and provide feedback here. I'm not immediately able to say yet what the implications of that will be for existing data and flows, especially if this PR continues to see more work done to it. Again, I'll report our findings here.

**Comment by RichardHeslopZA** [comment]
We have deployed this functionality into a QA environment and it appears to be working as expected (see screenshot below):

screenshot_2017-04-03-12-00-31

@SeanBlaschke Given the project urgency, we are going to build a fork containing this PR and deploy to your Production environment. Note that we're essentially building a fork on Production that is subject to active development by Pulilab. As the active development continues on the PR, Praekelt will need to invest time to re-release the build, QA it and then launch in production - we'll need to do this every time Pulilab does new work on this PR until it is merged into RapidPro core and released with a version number.

So, the pressure on completing this work (i.e. part of the Pulilab USSD implementation) is still on as much as ever.

**Comment by smn** [comment]
@nullcode nice work on the session closing work. I merged in the branch straight from your repository, I've not addressed any of the conflicting files mentioned above.

File: temba/channels/models.py [hunk 5]
```diff
@@ -1424,11 +1427,12 @@ def send_junebug_message(cls, channel, msg, text):
...
 
         if is_ussd:
-            external_id, session_status = Msg.objects.values_list('response_to__external_id', 'session__status').get(pk=msg.id)
+            session = USSDSession.objects.get_session_with_status_only(msg.session_id)
+            external_id = Msg.objects.values_list('external_id', flat=True).filter(pk=msg.response_to_id).first()
             # NOTE: Only one of `to` or `reply_to` may be specified
             payload['reply_to'] = external_id
             payload['channel_data'] = {
-                'continue_session': session_status not in ChannelSession.DONE,
+                'continue_session': session and not session.should_end,
```
Comment (smn) [inline temba/channels/models.py]: @nullcode there is a very subtle bug here: `session` can be `None` which would set `continue_session` to be `None` where only `True` or `False` are valid values.

Comment (smn) [inline temba/channels/models.py]: perhaps wrapping the `session` in a `bool()` would be enough

Comment (nullcode) [inline temba/channels/models.py]: @smn if it doesn't take None then we just have to include an or statement at the end, the question is what would be the default value ?
`session and not session.should_end or True`

**Comment by nullcode** [comment]
whoa what happened here, @nicpottier any idea why the tests blew up ?

File: temba/channels/models.py (hunk 5 above)
Comment (smn) [inline temba/channels/models.py]: I think it'll have to be `False` if `session` is `None`.

**Comment by nicpottier** [comment]
I'm guessing bad merge? Are you seeing this on your branch @smn?

**Comment by smn** [comment]
@nullcode & @nicpottier hm yes, seeing the same errors now: https://travis-ci.org/praekeltfoundation/rapidpro/builds/218921342

**Comment by smn** [comment]
git bisect doesn't give me much more than pointing at the 04dea1e6db42cdeead379c7b36585d1cb083b2b6 merge commit I'm afraid.

File: temba/channels/models.py (hunk 3 above)
Comment (nicpottier) [inline temba/channels/models.py]: It must have been covered before then, either the code path changed or the tests changed. So have to figure out how to get that back

**Comment by nicpottier** [comment]
Seems like a deep dive on one of these failures would probably root out the cause. master is passing so it has to be some interaction between this PR and changes, but ya looking through the diff again I don't see anything obvious. My guess was maybe something related to adding the `include_active` flag to `Flow.start` but that seems captured here.

**Comment by Reobone** [comment]
@nicpottier, do you perhaps know when the conflict will be resolved in the existing branch? We need to test after this pull request is merged that https://github.com/rapidpro/rapidpro/issues/528 is a persisting issue or not

**Comment by nicpottier** [comment]
@reobone We aren't working on this, I would talk to the PR owners or others interested in merging this.

**Comment by nullcode** [comment]
@nicpottier are you waiting for anything else than me fixing the tests here? I'm asking, because I fixed merge conflicts 5 times during this PR and it was still not merged after the reviews and a clean build. If tests are the only blocking issue here and I will spend some time figuring out which part of the internal api changed that I also used in this PR and fix them.

**Comment by ewheeler** [comment]
@nullcode i think there was a request for additional tests here: https://github.com/rapidpro/rapidpro/pull/423#discussion_r110672352

**Comment by nicpottier** [comment]
@nullcode I think now that Praekelt has some experience with this then we're cool with merging once tests are passing. We were hesitant to do so earlier (which I think was born out to be the right decision) since this code didn't have any real world testing.

**Comment by nullcode** [comment]
@nicpottier makes sense, I fixed the conflicts and tests now. I noticed that you guys won't allow the simulator to be used without a USSD channel added anymore, can you tell me the reasoning behind that ? I wanted to implement it so everyone can build USSD flows and try them out in the simulator without needing a channel.

**Comment by nicpottier** [comment]
@nullcode We thought it better to not muddy the channel session schema with a nullable channel.. would really like to avoid that if possible. Do agree it can be useful to create USSD flows without a channel, but given the few # of people who use USSD (vs say IVR) it seemed better to not introduce issues there.

**Comment by nullcode** [comment]
@nicpottier fair enough, I already reverted the nullable field and wrote tests to check that it can't be null so this is good to go.
